var/spack/repos/builtin/packages/elk/package.py

```diff
@@ -13,41 +13,83 @@ class Elk(MakefilePackage):
     homepage = 'http://elk.sourceforge.net/'
     url      = 'https://sourceforge.net/projects/elk/files/elk-3.3.17.tgz'
 
-    version('3.3.17', sha256='c9b87ae4ef367ed43afc2d43eb961745668e40670995e8e24c13db41b7e85d73')
-
-    # Elk provides these libraries, but allows you to specify your own
-    variant('blas',   default=True,
-            description='Build with custom BLAS library')
-    variant('lapack', default=True,
-            description='Build with custom LAPACK library')
-    variant('fft',    default=True,
-            description='Build with custom FFT library')
-
-    # Elk does not provide these libraries, but allows you to use them
-    variant('mpi',    default=True,
+    version('7.2.42', sha256='73f03776dbf9b2147bfcc5b7c062af5befa0944608f6fc4b6a1e590615400fc6')
+    version('7.1.14', sha256='7c2ff30f4b1d72d5dc116de9d70761f2c206700c69d85dd82a17a5a6374453d2')
+    version('7.0.12', sha256='9995387c681d0e5a9bd52cb274530b23c0370468b6be86f6c90a6ec445cb8a01')
+
+    # what linear algebra packages to use? the choices are
+    # internal - use internal libraries
+    # generic  - use spack-provided blas and lapack
+    # openblas - use openblas specifically, with special support for multithreading.
+    # mkl - use mkl specifically, with special support for multithreading
+    # should be used with fft=mkl
+    # blis - use internal lapack and blas implementation from blis
+    variant('linalg', default='internal', multi=False,
+            description='Build with custom BLAS library',
+            values=('internal', 'generic', 'openblas', 'mkl', 'blis'))
+    # what FFT package to use? The choices are
+    # internal - use internal library
+    # fftw - fftw3 with special code
+    # mkl  - use mklr with fft code
+    # should be used with linalg=mkls
+    variant('fft',    default='internal', multi=False,
+            description='Build with custom FFT library',
+            values=('internal', 'fftw', 'mkl'))
```

Comment:
Spack has a mechanism called virtual dependencies that may handle this better:
```console
$ spack providers fftw-api blas lapack
fftw-api:
amdfftw@2:  fftw@2.1.5  fujitsu-fftw     intel-oneapi-mkl                     intel-parallel-studio@composer.2017:
cray-fftw   fftw@3:     intel-mkl@2017:  intel-parallel-studio@cluster.2017:  intel-parallel-studio@professional.2017:

blas:
amdblis  atlas  blis         essl       fujitsu-ssl2  intel-oneapi-mkl       netlib-lapack  nvhpc     veclibfort
amdblis  blis   cray-libsci  flexiblas  intel-mkl     intel-parallel-studio  netlib-xblas   openblas

lapack:
amdlibflame  atlas  cray-libsci  flexiblas  fujitsu-ssl2  intel-mkl  intel-oneapi-mkl  intel-parallel-studio  libflame  netlib-lapack  nvhpc  openblas  veclibfort
```
Basically, in the package, you just say:
```python
depends_on('fftw-api')
depends_on('blas')
depends_on('lapack')
```
and users can specify which one to use by saying:
```console
$ spack install elk ^intel-mkl
```
Then in the package all you have to do is query `spec['fftw-api'].libs.ld_flags` and it will calculate the correct flags to use for that API implementation.

Reply:
Virtual dependencies on blas and lapack should be pulled by choosing 'generic' variants. Openblas, MKL, BLIS and FFTW require special handling, because elk has custom code to better integrate with those particular libraries, and relevant variants activate this code.

Reply:
In that case, you can still have a single dependency without a variant, and check which provider is being used like so:
```python
if self.spec['fftw-api'].name == 'fftw':
    ...
elif self.spec['fftw-api'].name in ['intel-mkl', 'intel-parallel-studio', 'intel-oneapi-mkl']:
    ...
else:
    ...
```

Reply:
I'm aware.  But what if the user wants to link with, say, MKL without activating optional integration code because of some limitations on multithreading in his particular setup? I tried to allow specification of as much options as possible here.

Reply:
How would a user do that with your current setup?

Reply:
select fft=mkl linalg=mkl to activate mkl integration. select linalg=generic ^intel-mkl to use mkl for linear algebra, but don't activate integration code.

I specifically disallowed use of fft=mkl without linalg=mkl  and vice versa because it doesn't appear to be supported by upstream.

Also, upstream doesn't mention support for external fft libraries except fftw and mkl, so I didn't use virtual dependency on fftw-api.